### DISCUSSION TITLE
Order of triggers of `before:response` intercept lifecyle is the same as the request lifecycle when having multiple interceptors of the same endpoint

### DISCUSSION THREAD

**Discussion by AronHoxha** [cypress-io/cypress#29733] [Questions and Help] [answered]
I noticed that when having multiple interceptors of the same api (lets say A, B, C)  and having a RouteHandler on each of them as follows
```js
    cy.intercept('<url>', req => {
      console.log('req A')
      req.on('before:response', res => {
        console.log('res A')
        // do something with response
      })
    })
```
The order of execution of the before:response handlers is the same as the outgoing request so:
req -> C -> B -> A
res -> C -> B -> A

![Screenshot from 2024-06-25 01-07-54](https://github.com/cypress-io/cypress/assets/93707789/7bd0a934-a3aa-4f62-9f6b-3f04bf87bfe7)

Whereas I would expect the order of the response to be in reverse for the response handler.
Is this the expected behavior (if not I can open an issue)?

Overall I am aiming to build a setup where you use multiple interceptors of a url to define a top to bottom structure of the final interceptor to be used in a test. So for example have a base interceptor in a beforeEach of a describe block and then intercept the same url in the test with the option to modify the response of the previous interceptor instead of creating a new one.
The above code snippet was how I envisioned it to work but due to the order being reversed it doesn't really work out. Is such a behavior achievable in another way?

Here is a full reproducible code snippet (note: the test is for angular component tests but I imagine the same behavior occurs in E2E and component tests of other frameworks):
```js
  it.only('should be a good example :D', () => {
    cy.intercept('http://localhost:4200/interesting-path', {first: 5})
    cy.intercept('http://localhost:4200/interesting-path', req => {
      console.log('req A')
      req.on('before:response', res => {
        console.log('res A')
        // do something with response
      })
    })
    cy.intercept('http://localhost:4200/interesting-path', req => {
      console.log('req B')
      req.on('before:response', res => {
        console.log('res B')
        // do something with response
      })
    })
    cy.intercept('http://localhost:4200/interesting-path', req => {
      console.log('req C')
      req.on('before:response', res => {
        console.log('res C')
        // do something with response
      })
    })

    cy.mount(AppComponent, {imports: [AppModule]})
      .then(() => fetch('http://localhost:4200/interesting-path'))
  });
```

**Comment by portyu9** [comment, marked as answer]
Hi @AronHoxha 

Yes — the important point is that Cypress does **not** treat multiple interceptors like nested middleware whose response callbacks automatically unwind in reverse order.

For normal intercepts, matching routes are processed in **reverse definition order**.

So if you register:

```js
cy.intercept('/api', handlerA)
cy.intercept('/api', handlerB)
cy.intercept('/api', handlerC)
```

the request-side matching order is:

```text
C
B
A
```

and you should not depend on the `before:response` listeners becoming:

```text
A
B
C
```

on the way back.

The current interception lifecycle processes all registered `before:response` listeners, then any `req.continue()` response callback, then `response` listeners. It is not modeled as a stack unwind.

If your goal is:

> define common interception behavior in `beforeEach`, then allow a test-specific intercept to override/extend it

Cypress already supports **request-route overriding** through reverse registration order. Define the common route first and the test-specific route later.

For middleware that must run before all ordinary matching routes, explicitly use:

```js
cy.intercept(
  {
    url: '/api/**',
    middleware: true,
  },
  (req) => {
    // common middleware
  }
)
```

`middleware: true` routes are processed first and in definition order.

For response transformation, I would avoid relying on relative registration order between several independent `before:response` handlers. Put transformations whose order matters into one controlled response pipeline instead:

```js
cy.intercept('/api/items', (req) => {
  req.continue((res) => {
    applyBaseTransform(res)
    applyTestSpecificTransform(res)
  })
})
```

There can only be **one `req.continue()` response callback per request**, so this also makes the intended ownership/order explicit.

Another option is to have the common helper accept a transformation callback:

```js
function interceptItems(transform = () => {}) {
  cy.intercept('/api/items', (req) => {
    req.continue((res) => {
      applyBaseTransform(res)
      transform(res)
    })
  })
}

interceptItems((res) => {
  res.body.name = 'overridden'
})
```

That is more deterministic than depending on undocumented listener ordering between several overlapping routes.

Please mark this answer as accepted if it helped, thank you.

**Reply by AronHoxha** [reply]
Thank you!
